### ISSUE TITLE
How to run on cpu ?

### ISSUE THREAD

**Issue by vipcualo** [TropComplique/FaceBoxes-tensorflow#5]
Thanks you !
I have only cpu and i run this project on cpy. My error : 
self._traceback = self._graph._extract_stack()  # pylint: disable=protected-access

InvalidArgumentError (see above for traceback): Default MaxPoolingOp only supports NHWC on device type CPU
	 [[Node: import/pool1/MaxPool = MaxPool[T=DT_FLOAT, data_format="NCHW", ksize=[1, 1, 3, 3], padding="SAME", strides=[1, 1, 2, 2], _device="/job:localhost/replica:0/task:0/device:CPU:0"](import/conv1/CRelu/Relu)]]
how to fix this error ?
Thanks you very much @TropComplique  ?

**Comment by vipcualo** [comment]
Can someone help me? Thanks

**Comment by TropComplique** [comment]
Hi. To run this on CPU you will need to change all `NCHW` tensors to `NHWC` tensors.

For example: 
1. Change axis here (`1 -> 3`):
https://github.com/TropComplique/FaceBoxes-tensorflow/blob/master/src/network.py#L22
2. Change `NCHW` to `NHWC` here:
https://github.com/TropComplique/FaceBoxes-tensorflow/blob/master/src/network.py#L36
3. Also, make changes in this function:
https://github.com/TropComplique/FaceBoxes-tensorflow/blob/master/src/detector.py#L242

But this is not all places. You can find other places by yourself.
After these changes, you will need to refreeze the graph using `save.py` and `create_pb.py`.

**Comment by tirtile** [comment]
https://pan.baidu.com/s/1Y5LJ8mzVKoALTEyaUklkAA  here is the cpu version of this PR, including a trained model.

**Comment by vipcualo** [comment]
@tirtile  do you convert trained model of TropCompile to CPU  or you trained from scratch  ?

**Comment by junedgar** [comment]
@tirtile  thank for your share. 
Can you give me advice to train your NHWC model?
I tried to train the faceboxes to detect bottle.
I trained the model by TropComplique's code, it  can detect the bottle. But when I used the same .tfrecord to train the model by your code, it detect nothing in same image. Is there anything change in generate the train dataset?
Looking forward for your reply!

**Comment by tirtile** [comment]
@vipcualo @junedgar  I trained the model according to config.json in this PR and changed nothing except data format.

**Comment by hahasofia** [comment]
@TropComplique @tirtile   thanks for your advises for changing NCHW to NHWC.
But I tired to change data format several times but still hard for me to fix it right. Could you share a complete CPU version for me? Thank you!

**Comment by alexliyang** [comment]
@tirtile ：when is it the loss value you stop the training?

**Comment by hahasofia** [comment]
@junedgar Do you find the reason? I used tritile's NHWC code also detected nothing in same image. It's very strange... Is there anything wrong during saving PB file?

**Comment by hahasofia** [comment]
@junedgar I used right model file, but I changed detector.py because I met error below when ran 'save.py'. So I'm not sure whether my changes affect the result. Do you meet below error? Thank you.
![image](https://user-images.githubusercontent.com/29223305/44379882-a6cdee00-a53a-11e8-8865-e185c29cb04e.png)

**Comment by ptrckqnln** [comment]
@tirtile I'd love to try your CPU model, but it looks like the link you posted expired. Any chance you can rehost it?

**Comment by TropComplique** [comment]
Hi. 
Now the `.pb` file also works on CPU. 
33 fps on an image with size 1024x768.

**Comment by junedgar** [comment]
@hahasofia  yes, I encountered this problem,  you can solve it by using python3.

**Comment by liuhaiyang2018** [comment]
@TropComplique 
Hi.
Use you provide .pb works on CPU speed:100ms per image, How did you test it?Is the test code
provider

**Comment by PBRAOS** [comment]
> @hahasofia yes, I encountered this problem, you can solve it by using python3.

@hahasofia , @junedgar I found a good fix on detector.py for python2 .
Look at the attachment.

Cheers,
pb
![screenshot from 2019-02-11 21-08-28](https://user-images.githubusercontent.com/6173454/52593654-65f7c780-2e41-11e9-9f3a-a54e39c645f4.png)

**Comment by marcoruizrueda** [comment]
> @TropComplique
> Hi.
> Use you provide .pb works on CPU speed:100ms per image, How did you test it?Is the test code
> provider

I got 300ms on CPU. Did you also get 33FPS?
